### ISSUE TITLE
The usefulness of tap interfaces without bridge

### ISSUE THREAD

**Issue by bictorv** [PDP-10/klh10#67]
I'm about to migrate my oldest TOPS-20 system from klh-2.0h (last touched 2012) to this version, and discovered that NI20 tap interfaces don't work as they used to. It seems there is a little bit of confusion about the usefulness of tap-only devices (as opposed to tap+bridge devices), given comments in the code. 

In my experience, tap-only devices are very useful as tunnel devices _if_ they have their IP addresses set up (as they were in 2.0h). I just committed [510aedc](https://github.com/PDP-10/klh10/commit/510aedca29b216e829dbe536cff6662234157bf6) which changes the crucial test to whether (in effect) the `tunaddr` parameter was given (indicating that it is to be used as a tunnel device) rather than checking `pfdata->pf_ip4_only` (indicating a `tun` (rather than `tap`) device).

I'm not sure if the other places in the code where tap+bridge is mentioned also need looking at.

And I think `OSN_USE_IPONLY` should be abolished in favor of `pfdata.pf_ip4_only`, but that's another matter.

(Sorry if this looks more like a rant than an issue.)

**Comment by Rhialto** [comment]
Looks good to me. I see you also adjusted the documentation. Nice.

I think it was this code in klh10 for setting up all those different networking scenarios for several operating systems that made me decide that from now on, I would not do it like that again. This stuff has lots of potential use cases, and is very non-portable, so it is easy to miss something. In the future I will leave *all* that sort of setup to the user. With a little script with the appropriate ifconfig and brconfig etc commands, you can set it up any way you want, even in ways that are difficult to handle when it's built into the virtual machine emulator. The documentation can mention some example cases that work for a system where I can test it, of course.

**Comment by bictorv** [comment]
I think usability is quite important, to support people who are not networking wizards, and to make it _easier_ to get things running. "Most people" are probably more interested in running the historic OS, rather than setting up the network in the host system. Perhaps a new option "noconf=yes" to turn such things off, if you really need to do things differently? "If it works, don't break it"?

People like you and me can always get by (although I must admit getting bridging to work on Linux seems a hassle without extensive examples and/or screen+keyboard - but for me, tap (with IP config) suffices.)

Maintaining support for n+1 different operating systems is a hassle, of course. In this particular case, I wonder if the "ifconfig" method for setting it up wouldn't work basically everywhere? Doing it "manually" with ioctls adds possibilities of headaches?

**Comment by Rhialto** [comment]
Yes especially the ioctl way is annoying. I guess that's why there are some cases where an external program is called to do a step of the work.

And you remind me of the time when I went to set up a bridging configuration on Linux (bridging a tap to eth0) and it broke the whole networking. After lots of searching it turned out that this is even by design. I don't know who thought of doing it that way, but they should never be let anywhere near a network interface ever again. So since then I have a permanent bridge configured, for the occasions where I want to try klh or simh or hercules or whatever. Just to avoid the stupid moving of the IP address from interface to bridge and back again.

I haven't looked in detail if a "noconf=yes" is needed; I seem to have managed without it so far. I think it autodetects if a new tun/tap needs to be created, or if it already exists. This of course also complicates the logic and if we overhaul it again, such an option could be useful to simplify.

But for the moment I'm in no hurry to change it.

**Comment by jayjwa** [comment]
SIMH's style (ATT XQ tap:tap3) where it just grabs the interface you tell it would be useful. QEMU will do this too (-net tap,ifname=tap8). I have 1-10 taps ready to go, all under br0, but KLH10 will not use one, even if told. It will always create a new one and leave it disconnected. As a work-around, I found that I can start TOPS20, and quickly move the new tap under the bridge while the system is coming up.
`ip -c link set dev tap11 master br0`

Previously I tried (Linux host) _tap+bridge_, thinking this would get me the desired behavior as described above, but this created a tap device and pasted an IP address on it. It was a week before I could find out why I could no longer connect to a news website, nypost.com (192.0.66.32), and several others. 192.168.20.x is my internal LAN. _tap+bridge_ had changed the routing table.

Currently this is working well,
`devdef ni0 564 ni20 dedic=true ifmeth=tap ifc=tap11 decnet=true`
other than the part of having to manually add the tap to the bridge (br0). In other words, it will not use the tap already available (tap10); it wants a new one each time, no matter what I specify for _ifc=_ . For clarity, I've set _ifc=tap11_ since this is the interface that will be created, even if tap10 is available and _ifc=tap10_ is specified. This is contrary to what's listed in install.txt.

> With both tun and tap you can specify with ifc=tun33 which of the tun/tap devices you want to use, for instance if you have one pre-set-up. Otherwise, one will be created for you.

`devdef ni0 564 ni20 dedic=true ifmeth=tap ifc=tap10 decnet=true`
yields:

```
[KNILDR: Loading microcode version 1(172) into Ethernet channel 0]

[dpni20: ifc "tap11" => ether f2:b:a4:75:b4:1a]
[dpni20:   VHOST 0.0.0.0]
[dpni20-W: "tap11" E/N addr changed: Old=f2:b:a4:75:b4:1a New=aa:0:4:0:13:4]
```

```
20: tap11: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast master br0 state UNKNOWN mode DEFAULT group default qlen 1000
    link/ether 72:80:14:96:26:77 brd ff:ff:ff:ff:ff:ff
```

**Comment by bictorv** [comment]
It looks like pfopen() in osdnet.c should (like the CENV_SYS_NETBSD case)  check if an explicit tunnel unit number is specified, and then just use that (with open()). Probably something like

```
    /* See if an explicit tunnel unit number is given */
    if (isdigit(osnpf->osnpf_ifnam[3])) {
#if CENV_SYS_NETBSD
	fd = pfopen_create(basename, tt_ctx, osnpf);
#else
	fd = open(osnpf->osnpf_ifnam, O_RDWR, 0);
#endif /* CENV_SYS_NETBSD */
	if (fd >= 0) {
	    return fd;
	}
    }
```

But I also think it's a separate issue?

**Comment by Rhialto** [comment]
It has been a long time since I wrote this networking stuff, so what @bictorv says sounds plausible but I didn't check it in depth.

Since then I have come to the conclusion that I simply should have left out as much as possible of this stuff from the emulator code base, and just ask the user to run the appropriate commands before starting emulation. The system dependencies and differences are just too horrible.

**Comment by bictorv** [comment]
@jayjwa did you try this?

**Comment by jayjwa** [comment]
I did try it, but it does not produce the desired result. It still wants to make its own tap, which won't be in the bridge unless you put it there, bypassing the already created one. 

```
[TOPS20 mounted]
AMATSU KLH, PANDA TOPS-20 Monitor 7.1(21733)-4 Internet: Loading host names [OK]

System restarting, wait...
Date and time is: Friday, 24-January-2025 12:42PM
Why reload? CM
Run CHECKD? n
 DDMP: Started
[KNILDR: Loading microcode version 1(172) into Ethernet channel 0]

[dpni20: ifc "tap11" => ether f2:b:a4:93:d1:2e]
[dpni20:   VHOST 0.0.0.0]
[dpni20-W: "tap11" E/N addr changed: Old=f2:b:a4:93:d1:2e New=aa:0:4:0:13:4]
24-Jan-2025 12:43:14 Internet: Network atr2net on, Output on

@info decnet
 Local DECNET node: AMATSU.  Nodes reachable: 1.
 Accessible DECNET nodes are:   AMATSU
```
Here you can see it's by itself, as the newly created tap isn't in the bridge. 

`devdef ni0 564 ni20 dedic=true ifmeth=tap ifc=tap10 decnet=true`

I did check to run the new binary.

**Comment by bictorv** [comment]
I suggest you add some debug printouts for the interesting code path in pfopen(), to see what happens.

**Comment by hdtodd** [comment]
On 2025.01.25, Jayjwa wrote:

> I did try it, but it does not produce the desired result. It still wants to make its own tap, which won't be in the bridge unless you put it there, bypassing the already created one.

It wasn't clear to me if requesting `... ifc=tap0` and getting `tap1` was intended behavior, but it seemed to complicate the setup of the network connections between DEC20 and host.

To understand the program flow in `osdnet.c`, I followed bictorv's suggestion and installed a number of `dbprintln()`'s and traced execution.  I found that in opening a cloning device, `pfopen()` corrupted `osnpf->osnpf_ifnam` just before returning -- the value of `pfname` hadn't been initialized.  That _seemed_ to be unintended.

Pull request #76 changes that behavior so that `pfname` is initialized at the beginning of `pfopen()`.  The result is that requesting `... ifc=tap0` gets you `tap0`, at least on Raspberry Pi running Debian Trixie.  I think that allows you to set up the bridge and tap in advance (temporarily via `ip` or permanently via `nmcli`) and simplifies the boot process.

I still am not able to get the DEC20+RPi system configured so that networking connections to the outside world work, and I'll continue to work on that.  But at least the network connection between the DEC20 guest and RPi host functions.

Thanks to you all for your continued work on this.  I hope this patch helps.

David

**Comment by jayjwa** [comment]
Excellent. This way, you don't have to shove the interface into the bridge after it makes a new one it doesn't need - it just uses the one you give it. This is what I meant and is similar to how the other emulators work. Seems to be working.

```
@login (USER) jayjwa (PASSWORD)                                                                                     
 Job 12 on TTY44 atr2.lan(TCP) 11-Dec-2025 18:28:17
  Last interactive login 9-Dec-2025 17:43:39
  Last non-interactive login 15-Sep-2025 18:14:08
 End of LOGIN.CMD.10
@info decnet
 Local DECNET node: AMATSU.  Nodes reachable: 7.
 Accessible DECNET nodes are:   AMATSU  ATR2    KIRIN   KUSHAL  LUNAST  NAMIEL
 TEOSTR
@info INTERNET
 Network atr2net host name is amatsu.lan
 Network interface type is IPNI, Internet address is 192.168.20.8
 Network interface is up, output is enabled
 Network service is enabled
 Last network interface up transition: 11-Dec-2025 18:26:03
@finger @mim.softjar.se
RSX-11M-PLUS system MIM. Fri Dec 12 00:28:56 2025. Up: 2 days, 20:40.

Luser         Real name         Term    Idle  Logged in     From
BILLQUIST     Johnny Billquist  TT10:     2d  09 Dec 03:48  50-73-179-7-pennsylv
BILLQUIST     Johnny Billquist  TT11:   7:36  09 Dec 03:49  50-73-179-7-pennsylv
@enable
$opr
OPR>enter ncp
NCP>loop node teostr
NCP>
18:43:02        NCP
                Request # 3 Accepted
NCP>
18:43:03        NCP

Request # 3; Loop Node Completed
NCP>
```
The config line and results:
```
devdef ni0 564 ni20 dedic=true ifmeth=tap ifc=tap10 decnet=true

ip -c link show tap10
16: tap10: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast master br0 state UP mode DEFAULT group default qlen 1000
    link/ether ca:fe:8c:fe:1d:9a brd ff:ff:ff:ff:ff:ff
```

**Comment by Rhialto** [comment]
I am on the road at the moment and can't easily try out this patch. It
looks like it should work and not break anything.

A quick look at the source made me think of 2 other possible strategies
to fix the issue. I didn't think them through fully, but I will mention
them below, just in case you think there might be some useful element in
them.

1. If CENV_SYS_NETBSD is set, there is a special function pfopen_create()
to deal with the case of a specified tap number. Maybe another
similar function for Linux would help.

2. The call to basenamecpy() at the end, which is the one corrupting
your result, could perhaps be moved into the loop which tries a sequence
of numbered tap names. Or in the else part of a if (fd < 0). So that it
only happens if pfname was certainly initialized.

That's it for now.

**Comment by hdtodd** [comment]
Thanks for the suggestions.  

I have little experience on other Linux/Unix systems, so I wanted to disturb as little of your code as possible.  That's why I stuck the initialization at the beginning.  Wouldn't do any harm, and did do some good.  😊 

But I like your suggestion #2 best.  Again, minimal code change (and unintended consequences) but same result.  I'll give it a try tomorrow and post the result here.

My "pull" request failed because the Github testing process couldn't find a "runner" to test with Ubunto 20.  Succeeded with 22 but couldn't even get tested with 20.

If my test with your suggestion #2 succeeds, I'll remove my prior pull request and resubmit the the new patch .

David

**Comment by hdtodd** [comment]
It appears that this patch also fixes networking for DEC20 running on a Mac OSX host.

**Comment by hdtodd** [comment]
Rhialto wrote last week regarding a code patch to avoid using uninitialized `pfname` in `pfopen()` in file `osdnet.c`: 

> 2. The call to basenamecpy() at the end, which is the one corrupting your result, could perhaps be moved into the loop which tries a sequence of numbered tap names. Or in the else part of a if (fd < 0). So that it only happens if pfname was certainly initialized.

Yes, I think your 2(b) is most straightforward.  I've applied the  patch below to my code (reversing the `if` condition for clarity), replacing the earlier patch, and confirmed that this works, too.  So I'll withdraw Pull #76 and replace it with a new pull request.

Either approach solves the problem; your suggested approach is cleaner and faster.

But a question: in the section `CENV_SYS_NETBSD`, if `fd>=0`, shouldn't `    tt_ctx->my_tap = TRUE;` before the return?  Seems like it's set TRUE in the other paths through the code if the open succeeds.

I've confirmed that this patch also works on both Mac and Raspberry Pi (Debian Trixie) in the same way the first patch worked:

- On the Mac, the guest DEC20 and host Mac **_can not_** interact via Internet, but both can interact with other computers on the local network.
- On the Pi, the guest DEC20 and host Pi **_can_** interact with each other via Internet, but neither can interact with other computers on the local network.

I'll continue to work on resolving those with my debugging version of `osdnet.c`.  

The patch:

```
--- osdnet.c.orig	2025-12-14 18:12:56
+++ osdnet.c	2025-12-18 11:00:06
@@ -1705,19 +1705,19 @@
 	    (void) snprintf(pfname, BASENAMESIZE, "%s%d", basename, i++);
 	    fd = open(pfname, O_RDWR, 0);
 	} while (fd < 0 && errno == EBUSY);	/* If device busy, keep looking */
-    }
-
-    if (fd < 0) {
-	/* Note possible error meanings:
-	   ENOENT - no such filename
-	   ENXIO  - not configured in kernel
-	*/
-	esfatal(1, "Couldn't find or open packetfilter device, last tried %s",
-		pfname);
-    }
+	if (fd >= 0) {
+            basenamecpy(osnpf->osnpf_ifnam, pfname, IFNAM_LEN);
+	} else {	  
+  	    /* Note possible error meanings:
+	       ENOENT - no such filename
+	       ENXIO  - not configured in kernel
+	    */
+	    esfatal(1, "Couldn't find or open packetfilter device, last tried %s",
+	 	       pfname);
+	};
+    };
 
     tt_ctx->my_tap = TRUE;
-    basenamecpy(osnpf->osnpf_ifnam, pfname, IFNAM_LEN);
 
     return fd;		/* Success! */
 }
```

Thanks for thinking this through.

David

**Comment by Rhialto** [comment]
Nicely done!

About this:
> But a question: in the section `CENV_SYS_NETBSD`, if `fd>=0`, shouldn't     `tt_ctx->my_tap = TRUE`; before the return? Seems like it's set `TRUE` in the other paths through the code if the open succeeds.

I looked at the function `pfopen_create()` and it is set in there in all cases, either `TRUE` or `FALSE`. So that's a valid worry but fortunately it is taken care of.

**Comment by hdtodd** [comment]
Rhialto,

Good!  Thanks for checking and for confirming that the additional 
|tt_ctx->my_tap = TRUE |statement isn't needed.

I see that the workflow check for my pull request (#77) has again been 
delayed (and will likely fail) because testing with Ubuntu 20 is required:.

    Requested labels: ubuntu-20.04
    Job defined at:
    ***@***.***/pull/77/merge
    Waiting for a runner to pick up this job...

It looks like github doesn't do Ubuntu 20 checks any more.  I don't know 
who the repository maintainers are, but the build workflow for pull 
requests should probably have that Ubuntu 20 requirement updated to 
Ubuntu 25.

And then could someone approve this pull request so it gets into the 
master repository?  I don't understand the process, but it would be good 
to get this patch into the master so that any new clones or forks get a 
corrected version of pfopen().

David

**Comment by Rhialto** [comment]
Hi David,

I made a change to the Ubuntu versions and from now on it should use 22 and 24. But that probably isn't active for your merge request yet.

On the other hand, your code seems to be part of the master branch already anyway. When I pushed my change I got some message about "Bypassed rule violations for refs/heads/master:" but it didn't actually stop me from doing it. Since github says you're a member of the organization, it probably allowed this for you too. I think I'll see if I can tweak some setting so that merge requests can't actually be circumvented (and also to make sure to keep a clean linear history where possible).
```
.../emulators/panda-dist/klh10$ git push
Enumerating objects: 9, done.
Counting objects: 100% (9/9), done.
Delta compression using up to 6 threads
Compressing objects: 100% (3/3), done.
Writing objects: 100% (5/5), 423 bytes | 423.00 KiB/s, done.
Total 5 (delta 2), reused 0 (delta 0), pack-reused 0 (from 0)
remote: Resolving deltas: 100% (2/2), completed with 2 local objects.
remote: Bypassed rule violations for refs/heads/master:
remote:
remote: - Changes must be made through a pull request.
remote:
To github.com:PDP-10/klh10.git
   23477d7..56001ed  master -> master
```

**Comment by Rhialto** [comment]
Hi @hdtodd I tried to find out how your commit ended up in the master branch already. First of all, it's fine, I think the code is ok and I was about to approve it anyway.

I was clicking around through various issues and merge requests and in this one https://github.com/PDP-10/klh10/pull/77/commits there is a last commit which merges something from a pull request to your own repo. In total the pull request claims it made no changes.  

I think there is some confusion and it arose because there is a way of doing things with github that isn't quite obvious when you start with it. Github requires that you fork a repository before you make a pull request, but it doesn't tell you that in your own fork, it is still best to *also* create a new branch for your proposed changes. This way, your fork remains "compatible" with any changes that happen in the original repository. (Any changes that are made in the original master branch can be pulled into the master branch of fork, and they won't be interfering with your own changes you're working on).

And you're not dealing with 2 branches called "master" which are located in different repositories. I suspect that you intended to push to (the master branch of) your own fork of the repository, but somehow ended up pushing to (the master branch of) the original. When working with different branch names, such a confusion becomes impossible.

However the workflow isn't obvious from the start, and is also not explained in many places, so it's not surprising people don't realise. I certainly didn't know about it at first, and it took me some time to realise why it's more or less necessary.

**Comment by hdtodd** [comment]
Hi, Rhialto,

Thanks for updating the Ubuntu version workflow.

OOPS!  I thought I had seen that my patch had already been merged but 
didn't think that would have been possible -- thought I was on my fork 
and misreading the message.  My apologies: I don't know how I became a 
member of the organization (it would be an honor that I don't deserve 
and shouldn't be trusted with!).

I have 4 installs of the code active right now (2 Macs and 2 Pis) as I 
consolidate work I've done on DEC20's over the last 20 years or so.  My 
first clone was directly from the master PDP-10/klh10 branch, and I must 
have been working in that clone when I did the "push".

I'll go back on those four systems to delete the clone and re-clone from 
my fork so that any future changes push to my fork rather than master.

Thanks for cleaning up the mess I created!

I'll continue to work on this to see if I can fix the problem with being 
unable to get DEC20 Internet access from outside the guest-host 
environment (Pi) or being able to only get Internet access to the guest 
from outside the host (Mac).  Then any changes I make will only push to 
my own fork, and I can generate pull requests from there.

David

**Comment by Rhialto** [comment]
Hi David,

other emulators, such as Open-SIMH, also seem to have networking issues on Macs. From what I remember, Apple has been making it hard or impossible for third parties to add  tun/tap devices to the kernel, or something like that. So it doesn't surprise me too much that klh10 also has networking issues in this context. I hope it can be fixed by "simply" finding the right setup for the networking stuff. It is a kind of "traditional" issue that networking based on bpf (or possibly even when using the pcap library) between the host and the emulator doesn't work (since bpf/pcap were never meant for that purpose when designed). That is what the tap + bridge construction is intended to fix. You may find inspiration by looking at other emulators, or qemu or that sort of programs.

(On Linux, bridges create other problems. Linux seems to think that if you add your real networking interface to a bridge, that it then should ignore all addresses on the interface. Instead it uses the addresses that are set on the bridge. If you didn't set any, this breaks your networking completely. I find this a ludicrous concept: bridges should not have addresses...)

**Comment by hdtodd** [comment]
Rhialto,

Thanks for the comments and insights.  They may help me work through 
some of the issues.

Yes, Apple seems determined to make sure no "foreign" (that is, 
non-Apple!) system infects their system.  I love my Mac because of its 
stability and integration; I hate it because it's so hard to do anything 
"different" on it (two sides of the same coin, I suppose).

I was delighted when I found that that patching the pfname issue fixed 
klh10 networking on the Mac -- except that the Mac can't talk to the 
DEC20!  I'm hoping that I can fix that by getting to know pcap a little 
better and fixing it.  It's easiest for me to work on DEC20 stuff from 
the Mac as that's my main tool for regular household work (email, 
surfing, correspondence, etc.). But I'm confident that, in the long 
term, Apple will make it impossible to run the DEC20 as a guest.  So I'm 
anxious to get klh10 networking on the Pi to be fully functional, and I 
can access the Pi-hosted DEC20 from the Mac to do DEC20 work.  Your 
comments about bridging on Linux may help me work through that.  I keep 
thinking I'm really close, but I'm still learning about how taps and 
bridges work, and the addressing and routing seem to be real issues.  
Your observations help.

I'm going to focus on the Pi (Debian) again to see if I can make a 
breakthrough.  Getting one of these two to be fully functional would be 
a great relief.  Thanks, again, for your support

**Comment by hdtodd** [comment]
I've been trying to get klh10 networking to be fully functional on a Raspberry Pi (Debian) Trixie system but have had only limited success.   After considerable reading, I _thought_ I understood the basic ideas of bridging and taps, but when I eventually coerced the network to look like I thought it should, the IP connectivity still wasn't working.

With the patch from pull #77 installed, the tap you ask for in the `klt20.ini` file is the tap you get.  With that patch installed:

1. On a fresh boot of the Pi, with just devices lo, eth0, and wlan0 defined, booting the DEC20 results in a system in which the Pi and DEC20 can communicate with each other over IP, but IP connectivity to the outside local network is broken for **both** systems.
2. On a fresh boot of the Pi, if `br0` and `tap0` are set up (below), booting the DEC20 results in a system in which there is **no** IP connectivity to/from either system.

So I could use some insights from those who really know how this should work.  If you see something that just looks wrong in the scripts or logs below, please let  me know.

I'm using `ip` to manage the network to stay within the NetManager environment that Debian seems to have moved to.  My local network is 192.168.1.0/24; my host Pi is .88 (eth0) and .89 (wlan0); I intend for the DEC20 to be .62 (specified in the klt20.ini file), so I designated .62 for the bridge device, `br0`.

I _think_ I should be specifying:

1. in `klt20.ini`, the DEC20 IP address is 192.168.1.62 
2. the bridge, `br0` in Linux, should have IP address 192.168.1.62
3. the tap, `tap0` in Linux, should have `br0` as its master but no IP address (but `dpni20` seems to assign one).

I set up the bridge and tap environment before booting `klh10`; `tap0` comes up when the DEC20 references it.  Since asking for `tap0` in `klt20.ini` now _gets_ `tap0` and not `tap1`, I can configure the bridge and tap, and move the tap under the bridge, from a Linux script _before_ booting the DEC20.  So it should all work smoothly.  But doesn't.

I think part of the problem may be the changes in the networking environment and tools as Linux has evolved.  For example, I'm suspicious of the `ifconfig` system command executed from `dpni20`:
```
[dpni20: osn_pfinit_tuntap, line 1993: execute 'ifconfig tap0 192.168.1.88 pointopoint 192.168.1.62 up']
```
since that seems to assign an address to `tap0` that I think it shouldn't have.  But removing that `ifconfig` and replacing it with a simple `ip link set tap0 up` didn't work, either.

To set up the bridge and tap, I adapted a script from Jayjwa:

```
% cat SetBrTap.bsh 
#!/bin/bash
#  'SetBrTap.bsh <bridge IP> <tap #>'
#  Set up bridge and tap to allow hosted TOPS20 system to use
#  Raspberry Pi's eth0 for Internet service
#  2025.12.31 HDTodd@gmail.com
#  Adapted from jayjwa's script,

set +x

# Color codes for fancy output.
RED='\033[0;31m'
GREEN='\033[0;32m'
BLUE='\033[0;34m'
NC='\033[0m'

if [ $# -ne 1 ]; then
   printf "\\n${RED}Script  to set up bridge and tap for 'klh10' Internet access${NC}\\n"
   printf "${RED}Usage './SetBrTap <tap #>'${NC}\\n"
   exit 1
fi

bridge_address="192.168.1.62/24"
bridge_name="br0"

printf "\\n${BLUE}Setting up bridge device '${bridge_name}' at address ${bridge_address}${NC}\\n\\n"

set -x

#----------------------------------------
#  Report our initial conditions before
#  setting up the bridge
printf "\\n${BLUE}Initial conditions${NC}\\n"
ip address show
ip route show
ip link show dev ${bridge_name}
ip tuntap show

#  Add the bridge, set its IP, and bring it up
printf "\\n${BLUE}Set up the bridge${NC}\\n"
ip link add name ${bridge_name} type bridge
ip link set ${bridge_name} type bridge stp_state 1
# If you're going to run DECnet, set the bridge lladdr here
ip link set ${bridge_name} up
ip -4 addr add ${bridge_address} broadcast + dev ${bridge_name}

#  Add the tap and bring it up under the bridge
printf "\\n${BLUE}Set up tap$1 ${NC}\\n"
ip tuntap add dev tap$1 mode tap
ip link set dev tap$1 master ${bridge_name}
ip link set tap$1 up
#? do we need to set routing for the bridge?
#ip r add via 192.168.1.1 proto static src ${bridge_address} default dev ${bridge_name}

#  Report final conditions
printf "\\n${GREEN}Final conditions${NC}\\n"
ip address show
ip route show
ip link show dev ${bridge_name}
ip tuntap show

exit 0
```

The network with DEC20 in operation looks like this:

```
$ ./show-ip.bsh
+ ip address show
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host noprefixroute 
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    link/ether 2c:cf:67:ad:c9:b9 brd ff:ff:ff:ff:ff:ff
    inet 192.168.1.88/24 brd 192.168.1.255 scope global dynamic noprefixroute eth0
       valid_lft 85979sec preferred_lft 85979sec
    inet6 fe80::c953:95e9:baf2:6a3a/64 scope link noprefixroute 
       valid_lft forever preferred_lft forever
3: wlan0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    link/ether 2c:cf:67:ad:c9:bb brd ff:ff:ff:ff:ff:ff
    inet 192.168.1.89/24 brd 192.168.1.255 scope global dynamic noprefixroute wlan0
       valid_lft 85980sec preferred_lft 85980sec
    inet6 fe80::5018:b779:1c7a:15a3/64 scope link noprefixroute 
       valid_lft forever preferred_lft forever
4: br0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether ae:93:5c:49:86:c9 brd ff:ff:ff:ff:ff:ff
    inet 192.168.1.62/24 brd 192.168.1.255 scope global br0
       valid_lft forever preferred_lft forever
    inet6 fe80::ac93:5cff:fe49:86c9/64 scope link tentative proto kernel_ll 
       valid_lft forever preferred_lft forever
5: tap0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel master br0 state UP group default qlen 1000
    link/ether ae:93:5c:49:86:c9 brd ff:ff:ff:ff:ff:ff
    inet 192.168.1.88 peer 192.168.1.62/24 brd 192.168.1.255 scope global tap0
       valid_lft forever preferred_lft forever
    inet6 fe80::ac93:5cff:fe49:86c9/64 scope link proto kernel_ll 
       valid_lft forever preferred_lft forever
+ ip route show
default via 192.168.1.1 dev eth0 proto dhcp src 192.168.1.88 metric 100 
default via 192.168.1.1 dev wlan0 proto dhcp src 192.168.1.89 metric 600 
192.168.1.0/24 dev br0 proto kernel scope link src 192.168.1.62 
192.168.1.0/24 dev tap0 proto kernel scope link src 192.168.1.88 
192.168.1.0/24 dev eth0 proto kernel scope link src 192.168.1.88 metric 100 
192.168.1.0/24 dev wlan0 proto kernel scope link src 192.168.1.89 metric 600 
+ ip link show dev br0
4: br0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP mode DEFAULT group default qlen 1000
    link/ether ae:93:5c:49:86:c9 brd ff:ff:ff:ff:ff:ff
+ ip tuntap show
tap0: tap persist
```
The DEC20 `info inter` shows that it thinks the network is up and that it has the .62 address.  But the Pi and the DEC20 can't communicate, and IP connections to/from the local network to the Pi don't work, either.

**Comment by hdtodd** [comment]
I suspended work trying to get bridge/tap working and went to pcap for both Mac and Pi, and I have them both working now.  I'll go back to trying bridge/tap but wanted to get *something* going.  

Thanks to you all for you encouragement, support, and ideas.

I wrote [this guide](https://github.com/hdtodd/DECSYSTEM-20-on-Raspberry-Pi) to help anyone who wants to run klh10/Panda on a Mac or Pi.  Corrections and suggestions would be welcome.

Ah, yes, I'd like to contribute this guide to _this_ site if that would be OK with the maintainers.  It has little value being off on its own rather than tied to this repository.  If it's OK, how might I do that?

Thanks again.

David

**Comment by larsbrinkhoff** [comment]
You could format your guide as a plain text file, or leave it as markdown, and submit a pull request.  It should go in the doc directory of klkh10.  Are you familiar with git?

**Comment by Rhialto** [comment]
This guide is great! Reading though the compilation instructions reminded me that the `configure` script is not supplied in the klh repository, but generated via `autogen.sh`. This is fairly normal in cases where formal releases are made, which then get to include a `configure` script. But that doesn't help you if you just build directly from the repository.

So I think it might be good to include the `configure` script that I happen to have around here. See https://github.com/PDP-10/klh10/pull/79 . I think it may differ slightly if different versions of autoconf and/or automake are used, so it would likely come out the same as for you (I used " GNU Autoconf 2.71").

This would simplify the building instructions: you don't need to install `autoconf`, and instead of running `../autogen...` it will be `../configure ...`.

**Comment by hdtodd** [comment]
On 1/11/26 8:44 AM, Rhialto The M. wrote:
> *Rhialto* left a comment (PDP-10/klh10#67) 
> <https://github.com/PDP-10/klh10/issues/67#issuecomment-3734574295>
>
> This guide is great! Reading though the compilation instructions 
> reminded me that the |configure| script is not supplied in the klh 
> repository, but generated via |autogen.sh|. This is fairly normal in 
> cases where formal releases are made, which then get to include a 
> |configure| script. But that doesn't help you if you just build 
> directly from the repository.
>
> So I think it might be good to include the |configure| script that I 
> happen to have around here. See #79 
> <https://github.com/PDP-10/klh10/pull/79> . I think it may differ 
> slightly if different versions of autoconf and/or automake are used, 
> so it would likely come out the same as for you (I used " GNU Autoconf 
> 2.71").
>
> This would simplify the building instructions: you don't need to 
> install |autoconf|, and instead of running |../autogen...| it will be 
> |../configure ...|.
>
You're right, it would make it simpler.  But there are significant 
differences between the Mac and Pi (Debian) "configure" files.  So much 
of the install process is common between the two that I'd rather not 
have two documents and two "configure" files, so it seems that 
continuing to install and use "autogen" is the least complicated 
approach overall and might serve more reliably as systems evolve.


Or did I misunderstand something?

**Comment by Rhialto** [comment]
One version of the configure script is supposed to work on all Unix-like systems. I would bet that if you compared the first few lines of both the configure scripts you mention, that they give a different version of autoconf. And you can put it to the test by copying both configure scripts to the other system and try to run them there. That should work.

The version I created (and actually it was years ago, back in 2021) starts like
```
#! /bin/sh
# Guess values for system-dependent variables and create Makefiles.
# Generated by GNU Autoconf 2.71 for klh10 2.0l.
```
If you take the version from my pull request you potentially have a 3rd version to try on both your systems.

**Comment by hdtodd** [comment]
Rhialto,

Thanks for the suggestion.  I took a look at the long list of "diff"'s 
but didn't look at the start.  I'll do that.

**Comment by hdtodd** [comment]
Rhialto,

I created fresh clones of PDP-10/klh10 on both Mac Tahoe and Raspberry 
Pi (Debian) Trixie, copied your two files into the appropriate places, 
and did the builds.  The executable files on both platforms certainly 
differed from the executables I had created with autogen, but 
klh10/Panda seem to run just fine on both Mac and Pi.

I'll do a little more testing, then generate a review of your pull 
request.  Is that enough to get it incorporated?  Once it's in the 
mainstream, I'll update the install guide, copy it into the klh10/doc 
directory, and generate a pull request.

I'll suggest your approach first but leave the autogen method 
description as a backup in case anyone has problems on other systems.

Thanks for the simplification!

David

**Comment by Rhialto** [comment]
Hi David,

you can also compare the `src/config.h` and the various `bld-*/Makefile` files that are generated after running the various configure scripts. That should be the only output from running `configure`. (Just in case I'm forgetting something, you can check that with the timestamps of all files.) If the generated files are all the same, then the compiled executables should be equivalent. They are usually not the same since I think you get things like timestamps in them, so compiling exactly the same sources twice can easily get 2 executables with are not 100% the same as files, but which are the same if you know which bits to ignore.

-Rhialto/Olaf

**Comment by Rhialto** [comment]
@hdtodd  Oh I also mean to anwer your question: yes, a (positive) review makes a "merge" button appear which I can click to incorporate the merge request into the main branch.

**Comment by hdtodd** [comment]
Thanks, Olaf (/Rhialto).  On the Pi, I compared the Makefiles for 
bld-kl, and here are the diffs:

    < SRC = ../../src
    ---
     > SRC = /home/hdtodd/DEC20Sources/klh10/src
    33c33
    < KLH10_HOME ?= ${DESTDIR}${exec_prefix}/bin
    ---
     > KLH10_HOME ?= ${DESTDIR}/DEC20

So the changes are just in directory definition formats.

Then I did a "make clean", built it again, and compared the kn10-kl file 
with the one I made earlier in the day, with no other changes.  The 
binary files differ so your point about the differences being in the 
timestamps seems correct -- nothing else had changed.

So I think you analysis is exactly right, and I'm comfortable that the 
config file changes you've suggested work just fine and simplify the 
process.  Review forthcoming shortly.

Thanks, again, for simplifying the process.

David

**Comment by hdtodd** [comment]
I didn't know whether I should just merge or merge & rebase, so I left 
for you to actually do the merge.

THANKS!

**Comment by Rhialto** [comment]
I used "rebase & merge", and I think I set a setting somewhere that a "merge commit" is not added. That's to get a clean, more linear history, at least when this is at all possible.
Thanks for reviewing!

Now the question is, if this solves the issue as a whole (so that it could be closed). I think we're not there yet, right? Or am I wrong?

**Comment by hdtodd** [comment]
On 1/14/26 5:16 PM, Rhialto The M. wrote:
> *Rhialto* left a comment (PDP-10/klh10#67) 
> <https://github.com/PDP-10/klh10/issues/67#issuecomment-3751954036>
>
> I used "rebase & merge", and I think I set a setting somewhere that a 
> "merge commit" is not added. That's to get a clean, more linear 
> history, at least when this is at all possible.
> Thanks for reviewing!
>
> Now the question is, if this solves the issue as a whole (so that it 
> could be closed). I think we're not there yet, right? Or am I 
> wrong?Message ID: ***@***.***>
>

I just did a complete fresh clone, configure, compile, install and test 
on the Mac, and it works just fine.

So I think we're done with this one and you can close it.  What do you 
think is missing?

**Comment by Rhialto** [comment]
I'll ask @jayjwa @bictorv who were earlier in the discussion if they agree the recent changes solve their issue, or if something remains that could better be handled in a new issue.

**Comment by hdtodd** [comment]
> You could format your guide as a plain text file, or leave it as markdown, and submit a pull request. It should go in the doc directory of klkh10. Are you familiar with git?

Lars, I created a branch "add-install-guide" (which I'll delete later).  I'm suggesting that we add a new directory, "install guides", and put things like this in there.  KLH's /doc files somehow seem 'sacred' to me, and I'm reluctant to mix my .md files with his .txt files.

Also, I'd hope that others might add install guides for other platforms, over time, and that would be a good place to collect them.

Thoughts?  I haven't generated a pull request yet ... looking for feedback on this suggestion.  But you can check out the install guide for Mac and Pi in there.

**Comment by jayjwa** [comment]
With a clean build (Slackware, Linux 6.12.64, gcc 15.2.0) just now, it seems OK and addresses my concerns about having to manually move the tap yourself. This works more like simh and qemu do. My first new run saw DECnet...sleeping...but on a restart it's fine.

**Comment by hdtodd** [comment]
Thanks, Jayjwa.  I'll generate a pull request tomorrow when I'm a little more awake.  I appreciate the checkout.  

With the edit in pull request #77, the tap you ask for is the tap you get.  So you can set up the bridge and tap, either for that session (using ip) or permanently (using nmcli or nmgui) in a NetworkManager environment and not need to move the tap manually again (I THINK).  But since I never got bridge/tap to work (yet), that's just a guess.  But at least you can now get the tap # you ask for in the klt20.ini file.

**Comment by hdtodd** [comment]
> With a clean build (Slackware, Linux 6.12.64, gcc 15.2.0) just now, it seems OK and addresses my concerns about having to manually move the tap yourself. This works more like simh and qemu do. My first new run saw DECnet...sleeping...but on a restart it's fine

Jayjwa, OK for me to add Slackware, Linux 6.12.64, gcc 15.2.0 as a system the installation guide has been tried with?

**Comment by jayjwa** [comment]
I certainly have no objects to that, but note that Slackware on x86_64 is quite different from Pi and Mac OS. We don't have _apt-get_, nor is _sudo_ set up in that way. Most of the packages are already present because the official (?) way to install Slackware is _The Holy Full Install_. There's no "-dev" packages either (the package includes the whole enchilada). For those that use NetworkManager, it would do most of the network configuration. For those doing without, the action is via /etc/rc.d/rc.inet1 with the settings from rc.inet1.conf . After that point, the admin would have to decide on the pcap or bridge method, with the resulting answer deciding what to have NM do or what to place in rc.inet1.conf . 

Other than those differences, I think the guide applies but then outside of those differences it's stock Linux stuff anyway.

**Comment by hdtodd** [comment]
On 1/18/26 6:54 PM, Jason W. Austin wrote:
> *jayjwa* left a comment (PDP-10/klh10#67) 
> <https://github.com/PDP-10/klh10/issues/67#issuecomment-3765896904>
>
> I certainly have no objects to that, but note that Slackware on x86_64 
> is quite different from Pi and Mac OS. We don't have /apt-get/, nor is 
> /sudo/ set up in that way. Most of the packages are already present 
> because the official (?) way to install Slackware is /The Holy Full 
> Install/. There's no "-dev" packages either (the package includes the 
> whole enchilada). For those that use NetworkManager, it would do most 
> of the network configuration. For those doing without, the action is 
> via /etc/rc.d/rc.inet1 with the settings from rc.inet1.conf . After 
> that point, the admin would have to decide on the pcap or bridge 
> method, with the resulting answer deciding what to have NM do or what 
> to place in rc.inet1.conf .
>
> Other than those differences, I think the guide applies but then 
> outside of those differences it's stock Linux stuff anyway.
>
Jayjwa,

Thanks for the details.  Sounds like there are enough differences that 
the guide might be more confusing than helpful for experienced Slackware 
managers, so best not to suggest it as a guide for them.
